### PR TITLE
ListStore integration for Post list

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR integrates the new `ListStore` into the post list. I wanted to separate this into multiple PRs, but because the changes are all connected, my attempts at separating it didn't give satisfactory results. Either the changes separated are too small, doesn't give enough context or creates too much merge conflicts for little gain. Instead of swimming against the tide, I wanted to open this PR as is and only separate it if the reviewer asks me to do so. In exchange, I'll add a lot of line comments in the PR itself to make it easier to review this.

_I am going to make a certain bot angry, but oh well 🤷‍. Onto the changes.._

---

We have been unable to add filters or searching to post list for some time due to pagination issues it would have caused. I have been working in the FluxC land for some time implementing a set of components which not only avoids the pagination issues, but also provides an easy way to add filters, search, order by, order or pretty much anything you can think of. This PR is the first one that utilizes these new components and it's considered to be the v1 integration of it.

### Testing Instructions

* Test a self-hosted site as well as a .com one. You don't have to try all of this on both, but at least verify that basics work on both type of sites. 

* Don't forget to rotate the screen at times. There are some rotation issues from before, but this PR shouldn't add any new ones.

* Use the emulator's "Cellular" settings to try a slower connection. (or take a subway ride 🤷‍) You don't have to test everything on a slow connection, but after you test everything on a fast connection, you can try the basics for a different site or after logging out and logging in on a slower one. If you slow the connection after everything is loaded once, it'll still load pretty fast, so won't be as useful of a test.

* If you find any issues, if you can try out the production version (or `develop`) to see if you can replicate the issue (where applicable), it'd be very helpful. If you can replicate it, that means it's most likely a non-related issue, so open a new issue and ping me in it so I can have a look to see if I can fix it quickly in this branch. If you can't, please document it in this PR.

### Testing Steps

Since this touches everything in post list anything that affects it should be tested. Here is a non-comprehensive list: (only because I can't think of all of it)

* Do a fresh installation of the app (or logout/login) and visit the "Blog Posts" and check that it loads the posts.

_A note on fresh installations: This will be only time the posts are loaded from scratch. After this, only the ids and modified dates of the post will be fetched, which means your subsequent visits or pull to refresh will be a lot faster and more bandwidth efficient (+90% difference on both of them in my tests)_

* Tap on a post and change the title or description and go back. The post should be updated and should say "Local changes".

* Tap on a post, go to "Post settings" and add a featured image, tap on update button and check that it's reflected in the post list after a brief wait. (network request)

_It looks like this doesn't work if the change only happens locally. Check #8472 for more details._

* Tap on the floating button and create a draft by filling in title and/or content and going back. Check that it shows up in the list.

* Disable your network, tap on the floating button and create a local draft by filling in the title and/or content. Check that it shows up in the list.

* Re-enable your network and tap on publish on the remote draft you created 2 steps ago. Check that it's showing that it's uploading and it's refreshed after the upload succeeds.

* Repeat the above step for the local draft you created.

* If you have a site with more than 100 posts, use that and scroll as far as you can go. (or your as much as your patience allows) Check that more posts are loaded as you go and load more progress bar is visible at times. You can also overwrite the `pageSize` in `PostListDescriptor` in `PostListFragment` if you don't have a site with as many posts.

* Change a post from another client. (Calypso, iOS etc) Pull to refresh the list and check that the post is updated in the post list.

* Trash a post in the list and check that it's removed and a snackbar is shown to undo. Check that it's deleted in another client.

* Trash a post in the list and tap on the undo bar to check that it's re-added to the list and it's not deleted from remote. (verify on remote after waiting a minute)

* For bonus points, change the `PostListDescriptorForRestSite` or `PostListDescriptorForXmlRpcSite` by passing more parameters to it and try out different filters.

---

@malinajirka since you're leaving for a conference, I asked @kwonye for the review. If you do end up having time, you're welcome to review and/or test the PR, the more the merrier.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt [hunk 1]
```diff
@@ -186,75 +170,89 @@ class PostListAdapter(
...
+                onPostSelectedListener?.onPostSelected(post)
+            }
         }
     }
 
-    private fun showFeaturedImage(postId: Int, imgFeatured: ImageView) {
-        val imageUrl = featuredImageUrls.get(postId)
+    private fun showFeaturedImage(post: PostModel, imgFeatured: ImageView) {
+        var imageUrl: String? = null
+        if (post.featuredImageId != 0L) {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt]: These actions should be taken in the bg thread, an issue is created for this #8474.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt [hunk 2]
```diff
@@ -596,116 +557,6 @@ class PostListAdapter(
...
     }
 
+    private class LoadingViewHolder(view: View) : RecyclerView.ViewHolder(view)
     private class EndListViewHolder(view: View) : RecyclerView.ViewHolder(view)
-
-    /*
-     * called after the media (featured image) for a post has been downloaded - locate the post
-     * and set its featured image url to the passed url
-     */
-    fun mediaChanged(mediaModel: MediaModel) {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt]: This is moved to `PostListFragment.onMediaChanged`.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt [hunk 3]
```diff
@@ -596,116 +557,6 @@ class PostListAdapter(
...
-                } else {
-                    featuredImageUrls.remove(post.id)
-                }
-                notifyItemChanged(position)
-            }
-        }
-    }
-
-    @SuppressLint("StaticFieldLeak")
-    private inner class LoadPostsTask internal constructor(private val mLoadMode: LoadMode)
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt]: We no longer need this task as `ListStore` takes care of it for us.

File: WordPress/src/main/res/layout/post_cardview_skeleton.xml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<android.support.v7.widget.CardView
```
Comment (oguzkocer) [inline WordPress/src/main/res/layout/post_cardview_skeleton.xml]: We need to improve this skeleton layout and have a separate issue for it ##8475.

File: WordPress/src/main/java/org/wordpress/android/ui/uploads/PostEvents.java [hunk 5]
```diff
@@ -4,18 +4,18 @@
 
 public class PostEvents {
     public static class PostUploadStarted {
-        public final int mLocalBlogId;
+        public final PostModel post;
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/uploads/PostEvents.java]: We need to access more than the siteId to be able to refresh the relevant row in post list, that's why this change was implemented for both `PostUploadStarted` and `PostUploadCanceled`.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java [hunk 6]
```diff
@@ -260,49 +258,6 @@ public static String getFormattedDate(PostModel post) {
         }
     }
 
-    public static boolean postListsAreEqual(List<PostModel> lhs, List<PostModel> rhs) {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java]: We no longer need these helpers.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt [hunk 7]
```diff
@@ -77,25 +102,32 @@ class PostListFragment : Fragment(),
...
     private lateinit var site: SiteModel
 
     @Inject internal lateinit var siteStore: SiteStore
     @Inject internal lateinit var postStore: PostStore
+    @Inject internal lateinit var listStore: ListStore
     @Inject internal lateinit var dispatcher: Dispatcher
 
+    private var listManager: ListManager<PostModel>? = null
+    private var refreshListDataJob: Job? = null
+    private val listDescriptor: PostListDescriptor by lazy {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt]: If you want to try different filters as mentioned in the PR, this is the value you need to modify.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt [hunk 8]
```diff
@@ -132,13 +172,46 @@ class PostListFragment : Fragment(),
...
+    }
+
     override fun onDestroy() {
         EventBus.getDefault().unregister(this)
         dispatcher.unregister(this)
 
         super.onDestroy()
     }
 
+    override fun onResume() {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt]: The only change in this method is the removal of `loadPosts(LoadMode.IF_CHANGED)` as it's no longer necessary with `ListStore`. It's moved for better organization, that's why it's showing as many changes. Here is the version in the base branch for easy comparison: https://github.com/wordpress-mobile/WordPress-Android/blob/feature/list-store-integration-for-post-list-master/WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt#L234-L249

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt [hunk 9]
```diff
@@ -132,13 +172,46 @@ class PostListFragment : Fragment(),
...
+            val delayMs = resources.getInteger(R.integer.fab_animation_delay).toLong()
+            Handler().postDelayed({
+                if (isAdded) {
+                    AniUtils.scaleIn(fabView, AniUtils.Duration.MEDIUM)
+                }
+            }, delayMs)
+        }
+    }
+
+    override fun onDetach() {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt]: This method is unchanged, it's moved for better organization. Here is the version from base branch for easy comparison: https://github.com/wordpress-mobile/WordPress-Android/blob/feature/list-store-integration-for-post-list-master/WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt#L329-L337

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt [hunk 10]
```diff
@@ -389,14 +372,94 @@ class PostListFragment : Fragment(),
...
+                }
+            }
+        })
+
+        postIdForPostToBeDeleted = post.id
+        shouldCancelPendingDraftNotification = true
+        snackbar.show()
+    }
+
+    private fun showPublishConfirmationDialog(post: PostModel) {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt]: The only change is the `isAdded` check, here is the version from base branch for easy comparison: https://github.com/wordpress-mobile/WordPress-Android/blob/feature/list-store-integration-for-post-list-master/WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt#L494-L511

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt [hunk 11]
```diff
@@ -491,175 +556,104 @@ class PostListFragment : Fragment(),
...
-            }
-        } else if (event.causeOfChange is CauseOfOnPostChanged.DeletePost) {
-            if (event.isError) {
-                val message = getString(R.string.error_deleting_post)
-                ToastUtils.showToast(nonNullActivity, message, ToastUtils.Duration.SHORT)
-                loadPosts(LoadMode.IF_CHANGED)
-            }
+    @Subscribe(threadMode = ThreadMode.BACKGROUND)
+    @Suppress("unused")
+    fun onListItemsChanged(event: OnListItemsChanged) {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt]: Whenever there is a change for a specific "type", FluxC will emit this event. In this case, the type will represent the posts belonging to a specific site. We need to check if the type that's updated is the one we are interested in and update the views accordingly.

The only type we have in FluxC is for posts right now. However, once we have more types, this will be an important check. Also, if you switch the site, the "type" will actually change, so if a post for another site is updated, the list will not be refreshed, as expected.

**Comment by oguzkocer** [comment]
@kwonye @malinajirka I've added the line comments as mentioned in the PR to make it easier to review. Please let me know if anything is not clear.

Also, please note that I haven't added screenshots or gifs to this PR. I think it'd be better if we keep this PR code focused. Since we are merging it to a master branch, we can use the PR against `develop` more UI focused with design reviews and such. (design changes are pretty minor anyway)

File: WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt [hunk 12]
```diff
@@ -134,43 +116,45 @@ class PostListAdapter(
...
+            VIEW_TYPE_ENDLIST_INDICATOR -> {
+                val view = layoutInflater.inflate(R.layout.endlist_indicator, parent, false)
+                view.layoutParams.height = endlistIndicatorHeight
+                EndListViewHolder(view)
+            }
+            VIEW_TYPE_LOADING -> {
+                val view = layoutInflater.inflate(R.layout.post_cardview_skeleton, parent, false)
+                LoadingViewHolder(view)
+            }
+            else -> {
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt]: In order to follow Fail-Fast principle we may consider explicitly specifying VIEW_TYPE_POST and throw an exception in the else branch.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt [hunk 13]
```diff
@@ -186,75 +170,89 @@ class PostListAdapter(
...
-                val cleanPostTitle = StringEscapeUtils.unescapeHtml4(post.title)
-                holder.title.text = cleanPostTitle
-            } else {
-                holder.title.text = context.resources.getText(R.string.untitled_in_parentheses)
-            }
+        listManager?.getItem(position)?.let { post ->
+            val context = holder.itemView.context
 
-            var cleanPostExcerpt = PostUtils.getPostListExcerptFromPost(post)
+            if (holder is PostViewHolder) {
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt]: AFAIU the holder should always be an instance of PostViewHolder. We may consider throwing an exception in the else branch or remove the if check so it crashes asap.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt [hunk 14]
```diff
@@ -204,100 +267,46 @@ class PostListFragment : Fragment(),
...
-     */
-    fun onEventMainThread(event: PostEvents.PostUploadCanceled) {
-        if (isAdded && site.id == event.localSiteId) {
-            loadPosts(LoadMode.FORCED)
+    private fun updateEmptyViewForListManagerChange(listManager: ListManager<PostModel>) {
+        if (!listManager.isFetchingFirstPage) {
+            if (listManager.size == 0) {
+                if (NetworkUtils.isNetworkAvailable(nonNullActivity)) {
+                    updateEmptyView(EmptyViewMessageType.NO_CONTENT)
+                }
```
[the remark this answers is not in the document]
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt]: This is actually a small workaround I added, because the empty view would briefly show the no content view when we are actually fetching the first page. Having said that, I am not even 100% sure about that claim, as I might be getting confused in a sea of issues I dealt with. This certainly requires at least a comment or an improvement before this PR is merged in.

Great catch, thanks @malinajirka!

File: WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt (hunk 12 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt]: Agreed, fixed in 8719cdf4fd7a263d114ee14e44f6402ffb8fec85.

The changeset looks bigger than it actually is in that commit due to indentation change. 😢 

That commit also updates the FluxC hash. It includes some important changes made in FluxC's `develop`  branch. If you already installed the app using this branch, you'll need to do a clean installation as the migration will fail. (had to change it due to a merge conflict) /cc @kwonye

File: WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt (hunk 13 above)
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt]: Agreed, fixed in 8719cdf4fd7a263d114ee14e44f6402ffb8fec85.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt]: @malinajirka I've improved the empty view handling in ee5e054544a044896ebbbbd749b2ccc49656a516. It should be much clearer now.

It's still not perfect though. I think ideally we'd want the `ListManager` to carry that information and the empty view is updated only through `ListManager` changes. I do have some plans for that, but not right now.

**Comment by oguzkocer** [comment]
Thanks for the review @malinajirka! I have addressed the issues you have raised.

> Only thing I think we may want to consider is moving some logic from the PostListFragment to separate classes. The fragment has almost 1k lines and it's quite difficult to follow. Since we are making a lot of changes anyway I think it would make sense to try to lighten it a bit.

I fully agree with your sentiment. Not only the fragment is too large, everything is intertwined at the moment. We need to move the logic to a view model, which will make this class much smaller. I gave this a shot yesterday night, but it's a significant undertaking and I don't think it'll happen, at least not before the merge.
